utils_cv/detection/model.py

```diff
@@ -285,63 +334,122 @@ def plot_precision_loss_curves(
 
     def evaluate(self, dl: DataLoader = None) -> CocoEvaluator:
-        """ eval code on validation/test set and saves the evaluation results in self.results. """
+        """ eval code on validation/test set and saves the evaluation results
+        in self.results.
+
+        Raises:
+            Exception: if both `dl` and `self.dataset` are None.
+        """
         if dl is None:
+            if not self.dataset:
+                raise Exception("No dataset provided for evaluation")
             dl = self.dataset.test_dl
         self.results = evaluate(self.model, dl, device=self.device)
         return self.results
 
+    def _transform(
+        self,
+        im: Union[str, Path, Image.Image, np.ndarray],
+    ) -> torch.Tensor:
+        """ Convert the image to the format required by the model. """
+        transform = transforms.Compose([transforms.ToTensor()])
+        im = transform(im)
+        if self.device:
+            im = im.to(self.device)
+        return im
+
+    @classmethod
+    def _apply_threshold(
+        cls,
+        pred: Dict,
+        threshold: Optional[int] = 0.5,
+        mask_threshold: Optional[int] = 0.5,
+        **kwargs,
+    ) -> Dict:
+        """ Return prediction results that are above the threshold if any. """
+        # detach prediction results to cpu
+        pred = {k: v.detach().cpu().numpy() for k, v in pred.items()}
+        # apply score threshold
+        if threshold:
+            selected = pred['scores'] > threshold
+            pred = {k: v[selected] for k, v in pred.items()}
+        if "masks" in pred and mask_threshold:
+            pred["masks"] = pred["masks"] > mask_threshold
+        return pred
+
+    def _get_det_bboxes_and_mask(
+        self,
+        pred: Dict,
+        im_path: Union[str, Path]
+    ) -> Dict:
+        """ Extract bounding boxes and masks from the prediction object. """
+
+        label_names = np.array(self.labels)[pred['labels'] - 1]
+        res = {
+            "det_bboxes":
+                DetectionBbox.from_arrays(
+                    pred['boxes'].tolist(),
+                    score=pred['scores'].tolist(),
+                    label_idx=pred['labels'].tolist(),
+                    label_name=label_names.tolist(),
+                    im_path=im_path,
+                )
+        }
+
+        if "masks" in pred:
+            res["masks"] = pred["masks"].squeeze(1)
+        return res
+
     def predict(
         self,
         im_or_path: Union[np.ndarray, Union[str, Path]],
         threshold: Optional[int] = 0.5,
-    ) -> List[DetectionBbox]:
+        **kwargs,
+    ) -> Dict:
         """ Performs inferencing on an image path or image.
 
         Args:
-            im_or_path: the image array which you can get from `Image.open(path)` OR a
-            image path
+            im_or_path: the image array which you can get from
+                `Image.open(path)` OR a image path
```

Comment:
[minor] OR should probably be lower case: "or an image path"

Reply:
corrected.